### ISSUE TITLE
Recommendations for users

### ISSUE THREAD

**Issue by stevejburge** [OSTraining/OSCampus#253]
We know what classes users have taken.

How do we use that to successfully suggest what classes to take next?

**Comment by htmgarcia** [comment]
We can look the pattern. So, if the user took a Joomla class, let's recommend to him another J! class from the same difficult level, or even recommend a class from the same pathway.

**Comment by nicksavov** [comment]
I don't think it should primarily be based on what classes they have already taken, which would require require Amazon-like algorithms and would be a big undertaking with a low ROI.

Instead, I think the focus should be on what the user's goals are. One thing that I had envisioned for OSCampus is the ability to do an initial training survey to customize the training for the user.

So for example, they sign up and immediately get redirected to the welcome video. Below it is the survey that asks questions one at a time (10 questions total at most) to custom the training for the user.

1) Which of these best describes you:
- I want to be a professional web developer and am committed to learning
- I want to be a professional web builder and am committed to learning. I’m only interested in learning the basics of coding to get me by, but I don’t want to be a developer.
- My company wants me to manage our website
- My company wants me to create a new website
- I don’t know anything about web development, but I’m exploring it as a potential career option
- I know exactly which class I’m looking for and just need to be pointed to the search
- I’m a student and want to learn on the side

And then based off of what they choose, you narrow it down with each question until you present them with a learning path

The above question is just an example to get the point across and it can be simplified to:

1) Which of these best describes you:
- I want to be a professional
- My company wants me to manage our website
- I know exactly which class I’m looking for and just need to be pointed to the search
- I’m a student and want to learn on the side
- I don’t know anything yet, and just want a simple introduction to creating websites

And then narrow it down with the next question.

The end result would be a recommended pathway. When they finish a class in the pathway, we could easily recommend the next class.

The pattern then is:
- Survey > pathway > class > next class in pathway > repeat until pathway is completed > start survey again.

**Comment by htmgarcia** [comment]
I like the idea of a survey, however it requires the user take it, and that means we depends on his actions (taking or not the survey).

If we can take advantage of the information we already have (i.e. which classes/pathways he took) and build a pattern of what the user is learning, that would automate the process and improve UX.

**Comment by nicksavov** [comment]
Yes, that's one short-coming of it (i.e. it requires the user to take the survey).

Yes, we could take advantage of the information we already have, but it will cost us dearly and we'll get very little return on it. I can't think of a simple way to recommend a new, relevant class based off classes that are already taken. But, maybe someone else can.

**Comment by htmgarcia** [comment]
New suggestion. When somebody finish a class, we should at least suggest to take the next class from the pathway.

**Comment by nicksavov** [comment]
+1, but the issue is which pathway? A class might be in many pathways.

That's why I think pathway has to come before class. And to go further, survey has to come before pathway.

**Comment by stevejburge** [comment]
We're not going to build an algorithm. There are tools that do that already:
http://getstream.io/features/
https://contextly.com/

**Comment by htmgarcia** [comment]
That's a good point, @nicksavov 

Let's see Netflix - after you watched a movie, their software suggest to watch other contents. So, in OSCampus, after passing the quiz, we can make suggestions related to current class (the concern is, how to determine the suggested class?) those tools may come in handy here @stevejburge 

I attach a quick draft:
![authority_and_analytics_quiz_-_search_engine_optimization](https://cloud.githubusercontent.com/assets/4999794/14570250/f13ee8c6-0308-11e6-861d-90f88bda2598.jpg)

**Comment by stevejburge** [comment]
@htmgarcia Yes, definitely. That's the idea here: https://github.com/OSTraining/OSCampus/issues/247

**Comment by nicksavov** [comment]
@stevejburge Nice finds! That would eliminate a lot of the initial work, but I doubt either of them would be good at finding relevant, personalized classes. Seems like it would be more useful for independent products based on other users' preferences (e.g. "Other shoppers also purchased"), rather than dependent (e.g. where they should go in order) products based on personal goals.

But, who knows, maybe students will like that approach more, rather than personalized, systematic training.

@htmgarcia , yes, it's definitely possible to do (e.g. Amazon and Netflix have done it), but my thinking is that it will be a ton of work and low ROI (since we don't have a lot of  users). We're a small company with limited resources and that's a big undertaking, IMO.

**Comment by stevejburge** [comment]
@nicksavov

http://getstream.io/  "BUILD SCALABLE NEWSFEEDS & ACTIVITY STREAMS IN A FEW HOURS INSTEAD OF WEEKS"

That's the promise, at least ...

**Comment by nicksavov** [comment]
I doubt it will work for what we need, but it's worth a try. I'd love it if I was wrong, btw. I don't have a horse in this race :D

**Comment by billtomczak** [comment]
When someone is viewing a lesson (regardless of type), there is a clear sequence from one class to the next is well defined and drives the prev/next buttons.

**completing a class**: We do not enforce any strict order of lessons. Which I believe is correct. But this means that *completing a class* could happen after watching a video or successfully passing a quiz in the middle of the class sequence. We don't currently provide any notification whatsoever that they just earned a certificate. This isn't good. Email should be sent, they should see a confirmation page. I do have a plan to correct this, but it will take some time.
Another issue with the concept of completion: Say someone successfully completes all the quizzes, but hasn't viewed all the videos? The instant they bring up the last of the unviewed videos - they don't have to watch it - the class is *complete*. At some point, when they move away from that video, we'll want to bring up a confirmation page. The means and rules for doing this are pretty sketchy in my mind.

**What's next?**: Although we don't enforce a strict order, we have prev/next buttons to indicate the next step. *Prev* does not appear for the first lesson, which seems fine. *Next* does not appear for the last lesson in a class. Which I think can be improved. Assuming *completing* the last lesson successfully means they've earned a certificate, that's one clear moment to display a confirmation page as well as suggestions for where to go next. Even if they do not deserve a certificate and completion page (didn't view all the videos, didn't pass an exam....) we can still provide suggestions based on a simple set of classes existing pathways, difficulty level and/or whatever. Something based on how we've organized pathways and classes. This would be pretty simple.

**recommendations**: I consider this a different and more complex problem having nothing to do with how we've structured things. @stevejburge's suggestion to use a 3rd party solution makes sense to me for this, although I'm not sure how getstream.io meets that requirement. As I recall, https://prediction.io/ looked more promising.

okay, probably already tl;dr - that's enough for now.

**Comment by stevejburge** [comment]
Should we start small? 

Start with the most basic completion page.

If we need to simplify our criteria for passing a class, that's fine by me.

Something > nothing in this case.

**Comment by billtomczak** [comment]
I think that's a first step regardless. We need a completion page and the *next class* recommendations I have in mind would be fairly simple to create for that. Sending email as part of showing the completion page is a minor additional detail.

**Comment by stevejburge** [comment]
Nice, thanks Bill

Yes, I think we do what we can. We're doing a good job of iterating.
